File: extract_death_places.js
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Extract frontmatter and death place information
function extractFrontmatter(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf-8');
    const match = content.match(/^---\n([\s\S]*?)\n---/);
    
    if (!match) return null;
    
    const yamlContent = match[1];
    const frontmatter = {};
    
    // Parse simple YAML lines
    yamlContent.split('\n').forEach(line => {
      if (line.includes(':')) {
        const [key, ...valueParts] = line.split(':');
        const value = valueParts.join(':').trim().replace(/^["']|["']$/g, '');
        frontmatter[key.trim()] = value;
      }
    });
    
    return frontmatter;
  } catch (error) {
    console.error(`Error reading ${filePath}:`, error.message);
    return null;
  }
}

function getConfidenceLevel(deathPlace) {
  if (!deathPlace || deathPlace === 'Research Needed') {
    return 'Low';
  }
  if (deathPlace.includes(',')) {
    return 'High';
  }
  return 'Medium';
}

function main() {
  const pioneersDir = path.join(__dirname, 'src', 'content', 'pioneers');
  
  // Get all markdown files
  const files = fs.readdirSync(pioneersDir)
    .filter(f => f.endsWith('.md') && !f.startsWith('1843-') && !f.startsWith('1850-') && !f.startsWith('early-'))
    .map(f => path.join(pioneersDir, f))
    .sort();
  
  console.log(`Found ${files.length} pioneer markdown files`);
  console.log('Extracting death place information...\n');
  
  const pioneersData = [];
  
  // Extract data from each file
  files.forEach(file => {
    const frontmatter = extractFrontmatter(file);
    
    if (frontmatter && frontmatter.name && frontmatter.birth && frontmatter.death) {
      const deathPlace = frontmatter.deathPlace || 'Research Needed';
      pioneersData.push({
        name: frontmatter.name,
        birth: frontmatter.birth,
        death: frontmatter.death,
        deathPlace: deathPlace,
        confidence: getConfidenceLevel(deathPlace)
      });
    }
  });
  
  // Sort by name
  pioneersData.sort((a, b) => a.name.localeCompare(b.name));
  
  // Print CSV header
  console.log('Name|Birth|Death|DeathPlace|Confidence');
  console.log('-'.repeat(88));
  
  // Print CSV data
  pioneersData.forEach(pioneer => {
    console.log(`${pioneer.name}|${pioneer.birth}|${pioneer.death}|${pioneer.deathPlace}|${pioneer.confidence}`);
  });
  
  // Print summary
  console.log('\n' + '='.repeat(88));
  console.log(`Total pioneers: ${pioneersData.length}`);
  
  const verified = pioneersData.filter(p => p.deathPlace !== 'Research Needed');
  console.log(`Verified death places: ${verified.length} (${(verified.length/pioneersData.length*100).toFixed(1)}%)`);
  
  const researchNeeded = pioneersData.filter(p => p.deathPlace === 'Research Needed');
  console.log(`Requiring research: ${researchNeeded.length} (${(researchNeeded.length/pioneersData.length*100).toFixed(1)}%)`);
  
  // Group by location
  const locations = {};
  verified.forEach(pioneer => {
    const loc = pioneer.deathPlace;
    if (!locations[loc]) {
      locations[loc] = [];
    }
    locations[loc].push(pioneer.name);
  });
  
  console.log('\nVerified Death Locations:');
  Object.keys(locations).sort().forEach(loc => {
    console.log(`  ${loc}: ${locations[loc].length} pioneers`);
  });
}

main();

File: extract_death_places.py
#!/usr/bin/env python3
"""
Extract death place information from all pioneer markdown files.
Outputs in pipe-delimited CSV format for database integration.
"""

import os
import re
import glob
from pathlib import Path
from typing import Dict, List, Tuple

# Script to extract death place information from markdown frontmatter
def extract_frontmatter(file_path: str) -> Dict:
    """Parse YAML frontmatter from markdown file."""
    frontmatter = {}
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
            
        # Match frontmatter between --- markers
        match = re.match(r'^---\n(.*?)\n---', content, re.DOTALL)
        if match:
            yaml_content = match.group(1)
            # Simple YAML parser for the key fields we need
            for line in yaml_content.split('\n'):
                if ':' in line:
                    key, value = line.split(':', 1)
                    key = key.strip()
                    value = value.strip().strip('"\'')
                    frontmatter[key] = value
        return frontmatter
    except Exception as e:
        print(f"Error reading {file_path}: {e}")
        return {}

def get_confidence_level(death_place: str) -> str:
    """Determine confidence level based on death place specificity."""
    if "Research Needed" in death_place or not death_place:
        return "Low"
    elif death_place and "," in death_place:  # Has location details
        return "High"
    else:
        return "Medium"

def main():
    """Main function to extract and process all pioneer death locations."""
    
    # Find all pioneer markdown files
    pioneers_dir = Path("src/content/pioneers")
    md_files = sorted(pioneers_dir.glob("*.md"))
    
    # Filter out non-pioneer files
    pioneer_files = [f for f in md_files if not f.name.startswith(("1843-", "1850-", "early-"))]
    
    pioneers_data: List[Dict] = []
    
    print(f"Found {len(pioneer_files)} pioneer markdown files")
    print("Extracting death place information...\n")
    
    # Extract data from each file
    for md_file in pioneer_files:
        frontmatter = extract_frontmatter(str(md_file))
        
        if frontmatter:
            name = frontmatter.get('name', '')
            birth = frontmatter.get('birth', '')
            death = frontmatter.get('death', '')
            death_place = frontmatter.get('deathPlace', 'Research Needed')
            
            if name and birth and death:
                confidence = get_confidence_level(death_place)
                pioneers_data.append({
                    'name': name,
                    'birth': birth,
                    'death': death,
                    'deathPlace': death_place,
                    'confidence': confidence
                })
    
    # Sort by name
    pioneers_data.sort(key=lambda x: x['name'])
    
    # Print CSV header
    print("Name|Birth|Death|DeathPlace|Confidence")
    print("-" * 88)
    
    # Print CSV data
    for pioneer in pioneers_data:
        csv_line = f"{pioneer['name']}|{pioneer['birth']}|{pioneer['death']}|{pioneer['deathPlace']}|{pioneer['confidence']}"
        print(csv_line)
    
    # Print summary statistics
    print("\n" + "=" * 88)
    print(f"Total pioneers: {len(pioneers_data)}")
    
    verified = [p for p in pioneers_data if p['deathPlace'] != 'Research Needed']
    print(f"Verified death places: {len(verified)} ({len(verified)/len(pioneers_data)*100:.1f}%)")
    
    research_needed = [p for p in pioneers_data if p['deathPlace'] == 'Research Needed']
    print(f"Requiring research: {len(research_needed)} ({len(research_needed)/len(pioneers_data)*100:.1f}%)")
    
    # Group by death location
    locations = {}
    for pioneer in verified:
        loc = pioneer['deathPlace']
        if loc not in locations:
            locations[loc] = []
        locations[loc].append(pioneer['name'])
    
    print("\nVerified Death Locations:")
    for loc in sorted(locations.keys()):
        print(f"  {loc}: {len(locations[loc])} pioneers")

if __name__ == "__main__":
    main()
